sdk/cosmos/src/headers/from_headers.rs

```diff
@@ -2,305 +2,208 @@ use crate::headers::*;
 }
 
-pub(crate) fn number_of_read_regions_from_headers(headers: &HeaderMap) -> Result<u32, AzureError> {
-    Ok(headers
-        .get(HEADER_NUMBER_OF_READ_REGIONS)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_NUMBER_OF_READ_REGIONS.to_owned()))?
-        .to_str()?
-        .parse()?)
+pub(crate) fn number_of_read_regions_from_headers(headers: &HeaderMap) -> Result<u32, HeaderError> {
+    parse(headers, HEADER_NUMBER_OF_READ_REGIONS)
 }
 
-pub(crate) fn activity_id_from_headers(headers: &HeaderMap) -> Result<uuid::Uuid, AzureError> {
-    let s = headers
-        .get(HEADER_ACTIVITY_ID)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_ACTIVITY_ID.to_owned()))?
-        .to_str()?;
-    Ok(uuid::Uuid::parse_str(s)?)
+pub(crate) fn activity_id_from_headers(headers: &HeaderMap) -> Result<uuid::Uuid, HeaderError> {
+    extract_with_parse(headers, HEADER_ACTIVITY_ID, |s| {
+        uuid::Uuid::parse_str(s).map_err(|e| HeaderError::ParsingError {
+            name: HEADER_ACTIVITY_ID.to_owned(),
+            error: e.into(),
+        })
+    })
 }
 
-pub(crate) fn content_path_from_headers(headers: &HeaderMap) -> Result<&str, AzureError> {
-    Ok(headers
-        .get(HEADER_CONTENT_PATH)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_CONTENT_PATH.to_owned()))?
-        .to_str()?)
+pub(crate) fn content_path_from_headers(headers: &HeaderMap) -> Result<&str, HeaderError> {
+    extract(headers, HEADER_CONTENT_PATH)
 }
 
-pub(crate) fn alt_content_path_from_headers(headers: &HeaderMap) -> Result<&str, AzureError> {
-    Ok(headers
-        .get(HEADER_ALT_CONTENT_PATH)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_ALT_CONTENT_PATH.to_owned()))?
-        .to_str()?)
+pub(crate) fn alt_content_path_from_headers(headers: &HeaderMap) -> Result<&str, HeaderError> {
+    extract(headers, HEADER_ALT_CONTENT_PATH)
 }
 
 pub(crate) fn resource_quota_from_headers(
     headers: &HeaderMap,
-) -> Result<Vec<ResourceQuota>, AzureError> {
-    let s = headers
-        .get(HEADER_RESOURCE_QUOTA)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_RESOURCE_QUOTA.to_owned()))?
-        .to_str()?;
-    Ok(resource_quotas_from_str(s)?)
+) -> Result<Vec<ResourceQuota>, HeaderError> {
+    extract_with_parse(headers, HEADER_RESOURCE_QUOTA, |s| {
+        resource_quotas_from_str(s).map_err(|e| HeaderError::ParsingError {
+            name: HEADER_RESOURCE_QUOTA.to_owned(),
+            error: e.into(),
+        })
+    })
 }
 
 pub(crate) fn resource_usage_from_headers(
     headers: &HeaderMap,
-) -> Result<Vec<ResourceQuota>, AzureError> {
-    let s = headers
-        .get(HEADER_RESOURCE_USAGE)
-        .ok_or_else(|| AzureError::HeaderNotFound(HEADER_RESOURCE_USAGE.to_owned()))?
-        .to_str()?;
-    Ok(resource_quotas_from_str(s)?)
+) -> Result<Vec<ResourceQuota>, HeaderError> {
+    let s = extract(headers, HEADER_RESOURCE_USAGE)?;
+    resource_quotas_from_str(s).map_err(|e| HeaderError::ParsingError {
```

Comment:
Nit: Any reason you aren't just using `extract_with_parse`? This seems almost the same as the function above

Reply:
Nope. I changed this before the `extract_with_parse` function existed. I'll change this.